File: src/hasoftware-library/src/hasoftware/api/messages/ErrorResponse.java
package hasoftware.api.messages;

import hasoftware.api.AnError;
import hasoftware.api.Message;
import hasoftware.cdef.CDEFMessage;
import hasoftware.cdef.CDEFSystemFlags;
import java.util.LinkedList;
import java.util.List;

public class ErrorResponse extends Message {

    private final List<AnError> _errorList;

    public ErrorResponse(int functionCode, int transactionNumber) {
        super(functionCode, transactionNumber, CDEFSystemFlags.Response | CDEFSystemFlags.Error);
        _errorList = new LinkedList<>();
    }

    public ErrorResponse(CDEFMessage cdefMessage) {
        super(cdefMessage);
        _errorList = new LinkedList<>();
        for (int count = cdefMessage.getU8(); count > 0; count--) {
            _errorList.add(new AnError(cdefMessage));
        }
    }

    public void addError(int number, int code, String message) {
        _errorList.add(new AnError(number, code, message));
    }

    public List<AnError> getErrors() {
        return _errorList;
    }

    @Override
    public void encode(CDEFMessage cdefMessage) {
        super.encode(cdefMessage);
        cdefMessage.putU8(_errorList.size());
        for (int count = 0; count < _errorList.size(); count++) {
            _errorList.get(count).encode(cdefMessage);
        }
    }
}

File: src.net/hasoftware-library/Messages/ErrorResponse.cs
using hasoftware.Api;
using hasoftware.Cdef;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace hasoftware.Messages
{
    public class ErrorResponse : Message
    {
        public List<AnError> Errors { get; private set; }

        public ErrorResponse(int functionCode, int transactionNumber)
            : base(functionCode, transactionNumber, CdefSystemFlags.Response | CdefSystemFlags.Error)
        {
            Errors = new List<AnError>();
        }

        public ErrorResponse(CdefMessage cdefMessage)
            : base(cdefMessage)
        {
            Errors = new List<AnError>();
            for (int count = cdefMessage.GetU8(); count > 0; count--) {
                Errors.Add(new AnError(cdefMessage));
            }
        }

        public void AddError(int number, int code, String message) {
            Errors.Add(new AnError(number, code, message));
        }

        public override void Encode(CdefMessage cdefMessage)
        {
            base.Encode(cdefMessage);
            cdefMessage.PutU8(Errors.Count);
            foreach (var error in Errors)
            {
                error.Encode(cdefMessage);
            }
        }
    }
}
